File: settlement/room/main/throne/InstanceConstruction.java
package settlement.room.main.throne;

import static settlement.main.SETT.ROOMS;
import static settlement.main.SETT.TWIDTH;

import init.resources.RESOURCE;
import init.sprite.SPRITES;
import settlement.job.ROOM_JOBBER;
import settlement.main.SETT;
import settlement.maintenance.ROOM_DEGRADER;
import settlement.path.AVAILABILITY;
import settlement.room.main.Room;
import settlement.room.main.TmpArea;
import settlement.tilemap.terrain.Terrain.TerrainTile;
import snake2d.Renderer;
import snake2d.util.color.COLOR;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.RECTANGLE;
import snake2d.util.datatypes.Rec;
import snake2d.util.sprite.SPRITE;
import util.colors.GCOLOR;
import util.rendering.RenderData;
import util.rendering.RenderData.RenderIterator;
import util.rendering.ShadowBatch;
import util.text.D;

final class InstanceConstruction extends Room.RoomInstanceImp implements ROOM_JOBBER{

	private static final long serialVersionUID = 1L;
	private final RECTANGLE body;
	private final int rot;
	private transient SPRITE icon;
	
	private boolean active;
	private int jobs;
	
	private static CharSequence ¤¤name = "Throne Construction";
	private static final int WORK = 8;
	static {
		D.ts(InstanceConstruction.class);
	}
	
	InstanceConstruction(int x1, int y1, int rot){
		super(SETT.ROOMS(), SETT.ROOMS().THRONE, false);
		
		THRONE p = SETT.ROOMS().THRONE;
		body = new Rec().moveX1Y1(x1, y1).setDim(Sprite.width(rot), Sprite.height(rot));
		if (SETT.ROOMS().map.get(p.construction) instanceof InstanceConstruction) {
			((InstanceConstruction)SETT.ROOMS().map.get(p.construction)).remove();
		}
		blueprintI().construction.set(body.cX(), body.cY());
		
		this.rot = rot;
		
		for (COORDINATE c : body) {
			setIndex(c.x(), c.y());
			SETT.ROOMS().data.set(this, c, 0);
		}
		SETT.ROOMS().map.init(this);
		jobs = area();
		active = !SETT.JOBS().planMode.is();
		for (COORDINATE c : body) {
			jobSet(c.x(), c.y(), active, null);
		}
		SETT.ROOMS().map.init(this);
	}
	
	@Override
	public boolean is(int tile) {
		return body.holdsPoint(tile%TWIDTH, tile/TWIDTH);
	}
	
	@Override
	public boolean is(int tx, int ty) {
		return body.holdsPoint(tx, ty);
	}
	@Override
	public int area() {
		return body().width()*body().height();
	}

	@Override
	public RECTANGLE body() {
		return body;
	}
	
	boolean active() {
		return body.width() > 0;
	}
	
	@Override
	protected boolean render(Renderer r, ShadowBatch shadowBatch, RenderData.RenderIterator it) {
		
		if (SETT.ROOMS().data.get(it.tile()) < WORK) {
			COLOR c = active ? GCOLOR.MAP().JOB_ACTIVE : GCOLOR.MAP().JOB_DORMANT;
			c.bind();
			SPRITES.cons().BIG.solid.render(r, 0, it.x(), it.y());
			COLOR.unbind();
		}
		return false;
	}
	
	@Override
	protected boolean renderBelow(Renderer r, ShadowBatch shadowBatch, RenderIterator i) {
		if (SETT.ROOMS().data.get(i.tile()) >= WORK) {
			blueprintI().sprite.renderFloor(r, shadowBatch, i);
		}
		return false;
	}
	
	public THRONE blueprintI() {
		return ROOMS().THRONE;
	}
	
	@Override
	protected AVAILABILITY getAvailability(int tile) {
		return AVAILABILITY.ROOM;
	}
	
	@Override
	public int mX() {
		return body.x1();
	}

	@Override
	public int mY() {
		return body.y1();
	}

	@Override
	public ROOM_DEGRADER degrader(int tx, int ty) {
		// TODO Auto-generated method stub
		return null;
	}

	@Override
	public void destroyTile(int tx, int ty) {
		// TODO Auto-generated method stub
		
	}

	@Override
	public boolean destroyTileCan(int tx, int ty) {
		return false;
	}

	@Override
	public CharSequence name(int tx, int ty) {
		return ¤¤name;
	}
	
	@Override
	public SPRITE icon() {
		if (icon == null)
			icon = new SPRITE.Twin(blueprintI().sprite.icon, SPRITES.icons().s.hammer);
		return icon;
	}

	@Override
	public int resAmount(int ri, int upgrade) {
		return 0;
	}

	@Override
	public void jobFinsih(int tx, int ty, RESOURCE r, int ram) {
		SETT.ROOMS().data.inc(this, tx, ty, 1);
		if (SETT.ROOMS().data.get(tx, ty) >= WORK) {
			jobs--;
			if (jobs == 0) {
				remove(body.x1(), body.y1(), false, this, false).clear();
				new Instance(body.x1(), body.y1(), rot);
			}
		}else {
			jobSet(tx, ty, active, null);
		}
	}

	@Override
	public void jobToggle(boolean toggle) {
		active = toggle;
	}

	@Override
	public boolean jobToggleIs() {
		return active;
	}

	@Override
	public boolean needsFertilityToBeCleared(int tx, int ty) {
		return true;
	}

	@Override
	public boolean becomesSolid(int tx, int ty) {
		return false;
	}

	@Override
	public int totalResourcesNeeded(int x, int y) {
		return 0;
	}

	@Override
	public TmpArea remove(int tx, int ty, boolean scatter, Object user, boolean forced) {
		for (COORDINATE c : body) {
			if (is(c)) {
				jobClear(c.x(), c.y());
			}
		}
		TmpArea t = super.delete(tx, ty, user);
		SETT.ROOMS().map.init(t);
		return t;
	}
	
	@Override
	public boolean needsTerrainToBeCleared(int tx, int ty) {
		TerrainTile t = SETT.TERRAIN().get(tx, ty);
		if (t.roofIs())
			return false;
		return ROOM_JOBBER.super.needsTerrainToBeCleared(tx, ty);
	}
	
	public void remove() {
		remove(mX(), mY(), false, this, true).clear();
	}

	@Override
	public boolean isJobActive() {
		return active;
	}
	

}

File: LegacySource/settlement/room/main/throne/InstanceConstruction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Snake2D;
using Snake2D.Util.Colors;
using Snake2D.Util.Datatypes;
using Snake2D.Util.Sprite;
using Util.Colors;
using Util.Rendering;
using Util.Text;

namespace Settlement.Room.Main.Throne
{
    internal class InstanceConstruction : Room.RoomInstanceImp, ROOM_JOBBER
    {
        private static readonly long SerialVersionUID = 1L;
        private readonly RECTANGLE body;
        private readonly int rot;
        private transient SPRITE icon;

        private bool active;
        private int jobs;

        private static readonly CharSequence ¤¤name = "Throne Construction";
        private const int WORK = 8;
        static
        {
            D.ts(typeof(InstanceConstruction));
        }

        public InstanceConstruction(int x1, int y1, int rot) : base(SETT.ROOMS(), SETT.ROOMS().THRONE, false)
        {
            THRONE p = SETT.ROOMS().THRONE;
            body = new Rec().MoveX1Y1(x1, y1).SetDim(Sprite.Width(rot), Sprite.Height(rot));
            if (SETT.ROOMS().Map.Get(p.construction) is InstanceConstruction)
            {
                ((InstanceConstruction)SETT.ROOMS().Map.Get(p.construction)).Remove();
            }
            BlueprintI().construction.Set(body.CX(), body.CY());

            this.rot = rot;

            foreach (COORDINATE c in body)
            {
                SetIndex(c.X, c.Y);
                SETT.ROOMS().Data.Set(this, c, 0);
            }
            SETT.ROOMS().Map.Init(this);
            jobs = Area();
            active = !SETT.JOBS().planMode.Is();
            foreach (COORDINATE c in body)
            {
                JobSet(c.X, c.Y, active, null);
            }
            SETT.ROOMS().Map.Init(this);
        }

        public override bool Is(int tile)
        {
            return body.HoldsPoint(tile % TWIDTH, tile / TWIDTH);
        }

        public override bool Is(int tx, int ty)
        {
            return body.HoldsPoint(tx, ty);
        }

        public override int Area()
        {
            return Body().Width() * Body().Height();
        }

        public override RECTANGLE Body()
        {
            return body;
        }

        private bool Active()
        {
            return body.Width() > 0;
        }

        protected override bool Render(Renderer r, ShadowBatch shadowBatch, RenderData.RenderIterator it)
        {
            if (SETT.ROOMS().Data.Get(it.Tile()) < WORK)
            {
                COLOR c = active ? GCOLOR.MAP().JOB_ACTIVE : GCOLOR.MAP().JOB_DORMANT;
                c.Bind();
                SPRITES.cons().BIG.solid.Render(r, 0, it.X, it.Y);
                COLOR.Unbind();
            }
            return false;
        }

        protected override bool RenderBelow(Renderer r, ShadowBatch shadowBatch, RenderIterator i)
        {
            if (SETT.ROOMS().Data.Get(i.Tile()) >= WORK)
            {
                BlueprintI().Sprite.RenderFloor(r, shadowBatch, i);
            }
            return false;
        }

        private THRONE BlueprintI()
        {
            return ROOMS().THRONE;
        }

        protected override AVAILABILITY GetAvailability(int tile)
        {
            return AVAILABILITY.ROOM;
        }

        public override int MX()
        {
            return body.X1();
        }

        public override int MY()
        {
            return body.Y1();
        }

        public override ROOM_DEGRADER Degrader(int tx, int ty)
        {
            return null;
        }

        public override void DestroyTile(int tx, int ty)
        {
        }

        public override bool DestroyTileCan(int tx, int ty)
        {
            return false;
        }

        public override CharSequence Name(int tx, int ty)
        {
            return ¤¤name;
        }

        public override SPRITE Icon()
        {
            if (icon == null)
                icon = new SPRITE.Twin(blueprintI().Sprite.Icon, SPRITES.icons().s.hammer);
            return icon;
        }

        public override int ResAmount(int ri, int upgrade)
        {
            return 0;
        }

        public override void JobFinsih(int tx, int ty, RESOURCE r, int ram)
        {
            SETT.ROOMS().Data.Inc(this, tx, ty, 1);
            if (SETT.ROOMS().Data.Get(tx, ty) >= WORK)
            {
                jobs--;
                if (jobs == 0)
                {
                    Remove(body.X1(), body.Y1(), false, this, false).Clear();
                    new Instance(body.X1(), body.Y1(), rot);
                }
            }
            else
            {
                JobSet(tx, ty, active, null);
            }
        }

        public override void JobToggle(bool toggle)
        {
            active = toggle;
        }

        public override bool JobToggleIs()
        {
            return active;
        }

        public override bool NeedsFertilityToBeCleared(int tx, int ty)
        {
            return true;
        }

        public override bool BecomesSolid(int tx, int ty)
        {
            return false;
        }

        public override int TotalResourcesNeeded(int x, int y)
        {
            return 0;
        }

        public override TmpArea Remove(int tx, int ty, bool scatter, object user, bool forced)
        {
            foreach (COORDINATE c in body)
            {
                if (Is(c))
                {
                    JobClear(c.X, c.Y);
                }
            }
            TmpArea t = base.Delete(tx, ty, user);
            SETT.ROOMS().Map.Init(t);
            return t;
        }

        public override bool NeedsTerrainToBeCleared(int tx, int ty)
        {
            TerrainTile t = SETT.TERRAIN().Get(tx, ty);
            if (t.RoofIs())
                return false;
            return ROOM_JOBBER.base.NeedsTerrainToBeCleared(tx, ty);
        }

        public void Remove()
        {
            Remove(MX(), MY(), false, this, true).Clear();
        }

        public override bool IsJobActive()
        {
            return active;
        }
    }
}